ggml WebGPU: add support for quantization types

reeselevine opened the pull request:
This PR adds support for basic matrix multiplication using many of the quantization types supported by ggml. Due to alignment requirements for WebGPU buffers, I have not yet implemented support for `mxfp4`, and WebGPU does not support `bf16` natively. Otherwise, all types that `test-backend-ops` are supported.


* Since WGSL compilers like Dawn do not support any preprocessing natively, the Python script for generating shaders does some minor preprocessing. I'm hopeful this will extend relatively easily to support more operations/more complex generation, but I'm open to other ideas if anyone has any.
* I noticed that some types, e.g., `Q8_1`, are not tested by `test-backend-ops`, so I didn't enable it in `supports_op`, and didn't implement other types not tested yet.
* It doesn't look like combinations of quantization and `fp16` are supported by the CPU backend, so I didn't enable those either.
* In the CI, I sometimes see failures that seem unrelated to my WebGPU implementation, mostly on Metal backends, where the result of the matrix multiplication is `NaN` on both the CPU and WebGPU side. For example (https://github.com/reeselevine/llama.cpp/actions/runs/17083427966/job/48442585422#step:7:26905). I'm not sure what the best solution is to this, I noticed that the Metal `test-backend-ops` doesn't actually seem to run any `MUL_MAT` operation tests at all.
* This PR also includes some other minor modifications, like adding some helper functions for tensor offset/alignment when binding buffers, and moving some initialization from `device_init` to `reg_get_device`, as I noticed this was necessary so that some objects are initialized by the time they are actually used.

reeselevine committed:
Begin work on set_rows

reeselevine committed:
Work on set rows

reeselevine committed:
Add error buffers for reporting unsupported SET_ROWS indices

reeselevine committed:
Remove extra comments

reeselevine committed:
Work on templating for different types in shaders

reeselevine committed:
Work on shader type generation

reeselevine committed:
Working q4_0 mul_mat and some templating for different types

reeselevine committed:
Add q4_0_f16 matmul and fix device init

reeselevine committed:
Add matmul support for basic quantization types

reeselevine committed:
Add q2_k and q3_k quantization

reeselevine committed:
Add rest of k-quants

reeselevine committed:
Get firt i-quant working

reeselevine committed:
Closer to supporting all i-quants

reeselevine committed:
Support rest of i-quants

reeselevine committed:
Merge remote-tracking branch 'origin/master' into types

reeselevine committed:
Cleanup code

reeselevine committed:
Merge pull request #2 from reeselevine/types

Types

reeselevine committed:
Fix python formatting

slaren commented:
> In the CI, I sometimes see failures that seem unrelated to my WebGPU implementation, mostly on Metal backends, where the result of the matrix multiplication is NaN on both the CPU and WebGPU side. 

> [NONE] NaN at index 136352 (WebGPU: WebGPU=nan CPU=nan) [MUL_MAT] NaN at index 64 (WebGPU: WebGPU=nan CPU=nan)   MUL_MAT(type_a=f16,type_b=f32,m=129,n=1,k=1057,bs=[1,1],nr=[1,1],per=[0,1,2,3],v=1): [1;31mFAIL[0m

The NaN present in a `NONE` node suggests that there was some issue initializing a tensor, this should never happen. If this consistently happens for f16, it might be an issue when converting f32 to f16 with this compiler. I would check that the `ggml_compute_fp16_to_fp32` function in `ggml-impl.h` works correctly with this compiler.

reeselevine committed:
debug

reeselevine committed:
Bugfix for memset

reeselevine committed:
Add padding to end of buffers on creation

reeselevine committed:
Simplify bit-shifting

reeselevine committed:
Merge pull request #3 from reeselevine/fixes

debug

reeselevine committed:
Update usage of StringView

reeselevine commented:
@slaren thanks for the investigation! Sorry I didn't find the issues earlier, I think I was misled by the `NaN` on the CPU result as well, and I'm still learning the codebase.

There was indeed a bug in the `memset` shader, I've verified that the updated code correctly sets the values of the last 1-3 bytes. I also updated the buffer allocation to align it to multiples of 4 bytes (WebGPU's binding size alignment), and fixed usages of `wgpu::StringView` to be converted to null-terminated strings when printed.

ggerganov reviewed (approved):
To improve the CI we can add a WebGPU node to [ggml-ci](https://github.com/ggml-org/llama.cpp/tree/master/ci).

reeselevine committed:
Merge remote-tracking branch 'upstream/master'
